# **Challenge: Quantitative Analysis: Case Study**

**Task:** to perform quantitative analysis to gain a deeper understanding of stock market dynamics and to inform investment strategies. The specific goals include:



1.   **Trend Analysis:** Identifying long-term trends in stock prices and market movements.
2.   **Volatility Assessment:** Evaluating the stability and risk associated with different stocks based on their price fluctuations.
3. **Correlation Study:** Investigating how different stocks correlate with each other, understanding market segments and diversification opportunities.
4.   **Risk-Return Trade-off Analysis:** Analyzing the balance between the potential risks and rewards of different stocks, aiding in portfolio management.














**Importing required dataset for analysis**

In [4]:
 from google.colab import files

uploaded=files.upload()

Saving stocks.csv to stocks (1).csv


In [5]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [6]:
stocks_data=pd.read_csv('stocks.csv')

In [7]:
# Checking loaded dataset

stocks_data.head()

Ticker        Date        Open        High         Low       Close  \
0   AAPL  2023-02-07  150.639999  155.229996  150.639999  154.649994   
1   AAPL  2023-02-08  153.880005  154.580002  151.169998  151.919998   
2   AAPL  2023-02-09  153.779999  154.330002  150.419998  150.869995   
3   AAPL  2023-02-10  149.460007  151.339996  149.220001  151.009995   
4   AAPL  2023-02-13  150.949997  154.259995  150.919998  153.850006   

    Adj Close    Volume  
0  154.414230  83322600  
1  151.688400  64120100  
2  150.639999  56007100  
3  151.009995  57450700  
4  153.850006  62199000

The dataset contains the following columns for stock market data:

Ticker: The stock ticker symbol.

Date: The trading date.

Open: The opening price of the stock for the day.

High: The highest price of the stock during the day.

Low: The lowest price of the stock during the day.

Close: The closing price of the stock for the day.

Adj Close: The adjusted closing price, which accounts for all corporate actions
such as dividends, stock splits, etc.

Volume: The number of shares traded during the day.

In [8]:
# Checking null values

stocks_data.isnull().sum()

Ticker       0
Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

In [9]:
# Checking datatypes of each column

stocks_data.dtypes

Ticker        object
Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

In [10]:
# Converting "Date" column datatype to datetime format

stocks_data['Date']=pd.to_datetime(stocks_data['Date'])

In [11]:
# Checking descriptive statistics of each column

stocks_data.describe()

Open        High         Low       Close   Adj Close  \
count  248.000000  248.000000  248.000000  248.000000  248.000000   
mean   215.252093  217.919662  212.697452  215.381674  215.362697   
std     91.691315   92.863023   90.147881   91.461989   91.454750   
min     89.540001   90.129997   88.860001   89.349998   89.349998   
25%    135.235004  137.440004  134.822495  136.347498  136.347498   
50%    208.764999  212.614998  208.184998  209.920006  209.920006   
75%    304.177505  307.565002  295.437500  303.942505  303.942505   
max    372.410004  373.829987  361.739990  366.829987  366.829987   

             Volume  
count  2.480000e+02  
mean   3.208210e+07  
std    2.233590e+07  
min    2.657900e+06  
25%    1.714180e+07  
50%    2.734000e+07  
75%    4.771772e+07  
max    1.133164e+08

In [12]:
# Let's check descriptive statistics for close prices

stock_stats=stocks_data.groupby('Ticker')['Close'].describe()
stock_stats

count        mean        std         min         25%         50%  \
Ticker                                                                     
AAPL     62.0  158.240645   7.360485  145.309998  152.077499  158.055000   
GOOG     62.0  100.631532   6.279464   89.349998   94.702501  102.759998   
MSFT     62.0  275.039839  17.676231  246.270004  258.742500  275.810013   
NFLX     62.0  327.614677  18.554419  292.760010  315.672493  325.600006   

               75%         max  
Ticker                          
AAPL    165.162506  173.570007  
GOOG    105.962503  109.459999  
MSFT    287.217506  310.649994  
NFLX    338.899994  366.829987

**Observations:**

1.   Average closing price for Apple is 158.24 and standard deviation is 7.36. The minimum closing price recorded is 145.39 while the maximum price the stock closed at is 174.  
2.   Average closing price for Google is 100.63 with lowest standard deviation of 6.27 as compared to others. Mninmum and maximum closing price recorded are 89.34 and 110
3.   Average closing price for Microsoft is 275 and standard deviation is 17.67. The minimum closing price recorded is 246.27 while the maximum price the stock closed at is 311.
4.  Average closing price for Netflix is 328 with highest standard deviation of 18.55 as compared to others indicating high variability and more stock price fluctuations . Mninmum and maximum closing price recorded are 293 and 367.





**Visualising the data**

In [13]:
# Pivoting the data for easy plotting

pivot_data = stocks_data.pivot(index='Date', columns='Ticker', values='Close')
pivot_data

Ticker            AAPL        GOOG        MSFT        NFLX
Date                                                      
2023-02-07  154.649994  108.040001  267.559998  362.950012
2023-02-08  151.919998  100.000000  266.730011  366.829987
2023-02-09  150.869995   95.459999  263.619995  362.500000
2023-02-10  151.009995   94.860001  263.100006  347.359985
2023-02-13  153.850006   95.000000  271.320007  358.570007
...                ...         ...         ...         ...
2023-05-01  169.589996  107.709999  305.559998  324.119995
2023-05-02  168.539993  105.980003  305.410004  317.549988
2023-05-03  167.449997  106.120003  304.399994  319.299988
2023-05-04  165.789993  105.209999  305.410004  320.779999
2023-05-05  173.570007  106.214996  310.649994  322.760010

[62 rows x 4 columns]

In [14]:
# Analysing the trend over time

fig=make_subplots(rows=1, cols=1)

for column in pivot_data.columns:
    fig.add_trace(
        go.Scatter(x=pivot_data.index, y=pivot_data[column], name=column), row=1, col=1)


fig.update_layout(title_text='Time series analysis of closing prices', xaxis_title='Date', yaxis_title='Closing Price',
                  legend_title='Ticker', showlegend=True,template='plotly_white')

fig.show()

**Observations:**



1.   Trend: Each stock shows its unique trend over time. For instance, AAPL and MSFT exhibit a general upward trend in this period.
2.   Volatility: There is noticeable volatility in the stock prices. For example, NFLX shows more pronounced fluctuations compared to others.
3. Comparative Performance: When comparing the stocks, MSFT and NFLX generally trade at higher price levels than AAPL and GOOG in this dataset.









In [15]:
# Visualising volatility of stocks

volatility_stats=pivot_data.std()
volatility_stats=volatility_stats.sort_values(ascending=False)
volatility_stats

Ticker
NFLX    18.554419
MSFT    17.676231
AAPL     7.360485
GOOG     6.279464
dtype: float64

In [16]:
fig=px.bar(volatility_stats, x=volatility_stats.index, y=volatility_stats.values,
           labels={'x': 'Ticker', 'y':'Standard Deviation'}, title='Volatility of closing prices')

fig.update_layout(template='plotly_white')

fig.show()

**Observations:**



1.   NFLX: Highest volatility with a standard deviation of approximately 18.55.
2.   MSFT: Next highest, with a standard deviation of around 17.68.
3.   AAPL: Lower volatility compared to NFLX and MSFT, with a standard deviation of about 7.36.
4.   GOOG: The least volatile in this set, with a standard deviation of approximately 6.28.

It indicates that NFLX and MSFT stocks were more prone to price fluctuations during this period compared to AAPL and GOOG.

In [17]:
# Let's analyse the correlation among stocks

stocks_correlation= pivot_data.corr()
stocks_correlation

Ticker      AAPL      GOOG      MSFT      NFLX
Ticker                                        
AAPL    1.000000  0.901662  0.953037  0.154418
GOOG    0.901662  1.000000  0.884527  0.201046
MSFT    0.953037  0.884527  1.000000  0.191273
NFLX    0.154418  0.201046  0.191273  1.000000

In [18]:
fig=go.Figure(data=go.Heatmap(z=stocks_correlation,x=stocks_correlation.columns,y=stocks_correlation.columns,
                              colorscale='blues', colorbar=dict(title='Correlation')))

fig.update_layout(title='Correlation Matrix of closing prices', xaxis_title='Ticker', yaxis_title='Ticker')



fig.show()

**Observation:**

AAPL and MSFT seem to have a relatively higher positive correlation.

In [19]:
pivot_data

Ticker            AAPL        GOOG        MSFT        NFLX
Date                                                      
2023-02-07  154.649994  108.040001  267.559998  362.950012
2023-02-08  151.919998  100.000000  266.730011  366.829987
2023-02-09  150.869995   95.459999  263.619995  362.500000
2023-02-10  151.009995   94.860001  263.100006  347.359985
2023-02-13  153.850006   95.000000  271.320007  358.570007
...                ...         ...         ...         ...
2023-05-01  169.589996  107.709999  305.559998  324.119995
2023-05-02  168.539993  105.980003  305.410004  317.549988
2023-05-03  167.449997  106.120003  304.399994  319.299988
2023-05-04  165.789993  105.209999  305.410004  320.779999
2023-05-05  173.570007  106.214996  310.649994  322.760010

[62 rows x 4 columns]

In [20]:
# Let's analyse the percentage change in closing prices from the start to the end of the period for each stock

percentage_change=((pivot_data.iloc[-1]-pivot_data.iloc[0])/pivot_data.iloc[0])* 100
percentage_change

Ticker
AAPL    12.234086
GOOG    -1.689193
MSFT    16.104798
NFLX   -11.073151
dtype: float64

In [21]:
fig=px.bar(percentage_change,x=percentage_change.index, y=percentage_change.values,
           labels={'y':'Percentage change (%)', 'x': 'Ticker'}, title='Percentage Change in Closing Prices')

fig.update_layout(template='plotly_white')

fig.show()

**Observation:**



1.   MSFT: The highest positive change of approximately 16.10%.
2.   AAPL: Exhibited a positive change of approximately 12.23% but lowered as compared to MSFT.
3. GOOG: Showed a slight negative change of about -1.69%. It indicates a minor decline in its stock price over the observed period.
4. NFLX: Experienced the most significant negative change, at approximately -11.07%. It suggests a notable decrease in its stock price during the period.





In [22]:
# Let's calculate the average daily return and the standard deviation of daily returns for each stock for risk-return analysis

daily_returns = pivot_data.pct_change().dropna()
daily_returns

Ticker          AAPL      GOOG      MSFT      NFLX
Date                                              
2023-02-08 -0.017653 -0.074417 -0.003102  0.010690
2023-02-09 -0.006912 -0.045400 -0.011660 -0.011804
2023-02-10  0.000928 -0.006285 -0.001972 -0.041766
2023-02-13  0.018807  0.001476  0.031243  0.032272
2023-02-14 -0.004225 -0.000526  0.003133  0.003876
...              ...       ...       ...       ...
2023-05-01 -0.000530 -0.004713 -0.005533 -0.017610
2023-05-02 -0.006191 -0.016062 -0.000491 -0.020270
2023-05-03 -0.006467  0.001321 -0.003307  0.005511
2023-05-04 -0.009913 -0.008575  0.003318  0.004635
2023-05-05  0.046927  0.009552  0.017157  0.006172

[61 rows x 4 columns]

In [26]:
avg_daily_return=daily_returns.mean()
risk= daily_returns.std()


In [27]:
risk_return=pd.DataFrame({'Risk': risk, 'Avg Daily Return': avg_daily_return})
risk_return

Risk  Avg Daily Return
Ticker                            
AAPL    0.014152          0.001991
GOOG    0.020710         -0.000067
MSFT    0.017881          0.002606
NFLX    0.022481         -0.001676

In [25]:
fig=go.Figure()

fig.add_trace(go.Scatter(x=risk_return['Risk'], y=risk_return['Avg Daily Return'], mode='markers+text',
                         text=risk_return.index, textposition='top center', marker_size=10 ))

fig.update_layout(title='Risk Vs. Return Analysis', xaxis_title='Risk (Standard Deviation)',
    yaxis_title='Average Daily Return', template='plotly_white')

fig.show()

**Observation:**



1.   AAPL shows the lowest risk combined with a positive average daily return, suggesting a more stable investment with consistent returns.
2.   GOOG has higher volatility than AAPL and, on average, a slightly negative daily return, indicating a riskier and less rewarding investment during this period.
3. MSFT shows moderate risk with the highest average daily return, suggesting a potentially more rewarding investment, although with higher volatility compared to AAPL.
4. NFLX exhibits the highest risk and a negative average daily return, indicating it was the most volatile and least rewarding investment among these stocks over the analyzed period.

# **Conclusion**

From the all four stocks observed on basis of closing prices trend, volatility, daily return percentages, Apple and Microsoft found to be most rewarding investment options with moderate to low volatility and positive daily average return. Other two, Netflix and Google could be rewarding in future if their volatility reduces.  